### ISSUE TITLE
How to determine the input layer size and optimize GPU utilization with Ultralytics YOLOv8?

### ISSUE THREAD

**Issue by HeavensDoor1** [ultralytics/ultralytics#18310]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi. I'm struggling with Ultralytics. I'm trying to fully utilize the GPU, but I notice that the higher the resolution, the slower it runs. I feed images of any size into the neural network, and it seems like Ultralytics resizes them internally. I've tried resizing the images myself before passing them to the network, but I can't figure out how to get the input size from the loaded model.

I trained the model with imgsz=768, using yolov8seg. The image is processed inside the network at a resolution of 448x768.

Here's how I loaded the model:
model = YOLO(model_path)
I want to extract the first input layer size: 448x768x3.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @HeavensDoor1, thank you for your interest in Ultralytics 🚀! We recommend checking out our [Docs](https://docs.ultralytics.com) where many common questions and usage examples are covered, including details on [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) workflows.

For your question about input size and GPU optimization:

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) that includes your code, the environment you're running in, and any relevant model file information to help us debug thoroughly.

If this is a ❓ Question, try to include as much supporting detail as possible, such as model configurations, any relevant training or inference logs, and the image resolutions you're working with.

In the meantime, we strongly suggest ensuring you're using the latest version of the `ultralytics` package, as improvements and optimizations are released regularly. Use the following command to upgrade:

```bash
pip install -U ultralytics
```

## Environments

YOLO models may be run in any of the following up-to-date verified environments, all with dependencies preinstalled:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

For troubleshooting GPU utilization or image resolution, please verify that your system meets the [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) and that all dependencies, including [CUDA](https://developer.nvidia.com/cuda), [CUDNN](https://developer.nvidia.com/cudnn), and [PyTorch](https://pytorch.org/), are correctly installed.

## Community Support

For real-time discussions or additional questions, join our community:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for live chat
- [Discourse](https://community.ultralytics.com) for in-depth discussions
- [Subreddit](https://reddit.com/r/Ultralytics) to collaborate with others

An Ultralytics engineer will review your issue and provide further assistance soon. Stay tuned! 😊

**Comment by Y-T-G** [comment]
Inference on a single image will very likely not fully utilize GPU. You can try inferring on a batch by passing a folder path and setting `batch` argument, or by passing a list of image path or numpy arrays.

It will still have I/O bottleneck where the GPU would be idle.

GPU utilization is also measured and averaged intervals. So your GPU might have reached 100% utilization for 1ms, but you will not be able to notice that because it was for such a short duration.

**Comment by Y-T-G** [comment]
What's your inference code?

**Comment by HeavensDoor1** [comment]
I made a small mistake, it's just a folder with images being processed, and it's visible how the processing becomes slower depending on the resolution of the input images. At the same time, as I understand it, the resizing is done by the CPU, and only after that it goes to GPU processing.
The code I have is literally like this:
model = YOLO(model_path)
results = model.predict(image, verbose=True)
I want to get the first input layer size for the next processed image: 448x768x3

**Comment by Y-T-G** [comment]
You can pass the folder path and set batch.

```python
results = model(folder_path, batch=64, half=True)
```

Batch inference will increase GPU utilization.

**Comment by HeavensDoor1** [comment]
Well, this will solve the GPU load issue, but some images are still processed quite slowly, and from the CPU load, it seems that this is most likely due to resizing. And again, is there a way to get the first input layer size of the image from YOLO?

**Comment by Y-T-G** [comment]
It's not the size of the first input layer. It's the size after LetterBox resizing with minimum rectangle padding.

You can check the code here:
https://github.com/ultralytics/ultralytics/blob/0504d3a8e004abf068d5c7b7bb3670602d6aa1ef/ultralytics/data/augment.py#L1475
